From 0b4d36b381d09682939bc413d539c2331bd5431f Mon Sep 17 00:00:00 2001
From: Matthew Lugg <mlugg@mlugg.co.uk>
Date: Sat, 15 Nov 2025 09:45:22 +0000
Subject: [PATCH 062/133] build.zig: use `Step.UpdateSourceFiles` for
 `zig1.wasm`
X-Developer-Signature: v=1; a=openpgp-sha256; l=1705; i=xtex@astrafall.org;
 h=from:subject; bh=6UnJT2R8O0TeGEqnViOCa2auKizm98hXIKRxfMp5hMM=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78pgD2QRUC+z2xP2ZosYXvWaZ8awalTUWJxccF1qaz
 jctY2Jrx0QWBjEuBksxRZYiwwZvVp10ftFl5bIwc1iZQIZIizQwMDAwsDDw5SbmlRrpGOmZahvq
 GRrqGOkYMXBxCsBUm7MzMhy8spYjwHlhIdOF/GNLlm9UCltZYd8sVHexI+fITE9zpX0M/wMid2p
 d79zHv3W61wT58yt/u4csUtI9/X9eqGeRc+17MwYA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

We were already using this for `stage1/zig.h`, but `stage1/zig1.wasm`
was being modified directly by the `wasm-opt` command. That's a bad idea
because it forces the build system to assume that `wasm-opt` has side
effects, so it is re-run every time you run `zig build update-zig1`,
i.e. it does not interact with the cache system correctly. It is much
better to create non-side-effecting `Run` steps (using `addOutput*Arg`)
where possible so that the build system has a more correct understanding
of the step graph.

(cherry picked from commit cd8fdd252d9ffd5bfb772cef49e4c79041872d38)
---
 build.zig | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/build.zig b/build.zig
index e834c9862ad0..11ae48413792 100644
--- a/build.zig
+++ b/build.zig
@@ -690,14 +690,14 @@ fn addWasiUpdateStep(b: *std.Build, version: [:0]const u8) !void {
     });
     run_opt.addArtifactArg(exe);
     run_opt.addArg("-o");
-    run_opt.addFileArg(b.path("stage1/zig1.wasm"));
+    const optimized_wasm = run_opt.addOutputFileArg("zig1.wasm");
 
-    const copy_zig_h = b.addUpdateSourceFiles();
-    copy_zig_h.addCopyFileToSource(b.path("lib/zig.h"), "stage1/zig.h");
+    const update_zig1 = b.addUpdateSourceFiles();
+    update_zig1.addCopyFileToSource(optimized_wasm, "stage1/zig1.wasm");
+    update_zig1.addCopyFileToSource(b.path("lib/zig.h"), "stage1/zig.h");
 
     const update_zig1_step = b.step("update-zig1", "Update stage1/zig1.wasm");
-    update_zig1_step.dependOn(&run_opt.step);
-    update_zig1_step.dependOn(&copy_zig_h.step);
+    update_zig1_step.dependOn(&update_zig1.step);
 }
 
 const AddCompilerModOptions = struct {
-- 
2.52.0

